source/connection_manager.c

```diff
@@ -1594,17 +1665,61 @@ static void s_cull_idle_connections(struct aws_http_connection_manager *manager)
     s_aws_http_connection_manager_execute_transaction(&work);
 }
 
+static void s_cull_pending_acquisitions(struct aws_http_connection_manager *manager) {
+    AWS_LOGF_INFO(AWS_LS_HTTP_CONNECTION_MANAGER, "id=%p: culling idle connections", (void *)manager);
+
+    if (manager == NULL || manager->pending_connections_acquire_timeout_ms == 0) {
+        return;
+    }
+
+    uint64_t now = 0;
+    if (manager->system_vtable->aws_high_res_clock_get_ticks(&now)) {
+        return;
+    }
+
+    struct aws_connection_management_transaction work;
+    s_aws_connection_management_transaction_init(&work, manager);
+
+    aws_mutex_lock(&manager->lock);
+
+    /* Only if we're not shutting down */
+    if (manager->state == AWS_HCMST_READY) {
+        const struct aws_linked_list_node *end = aws_linked_list_end(&manager->pending_acquisitions);
+        struct aws_linked_list_node *current_node = aws_linked_list_begin(&manager->pending_acquisitions);
+        while (current_node != end) {
```

Comment:
maybe a bit less copy/pasty to move the cull_acquisitions and cull_idle logic into the same function. just have them both put their actions in the same `work` struct

fewer locks, clocks, and if-we're-not-shutting-down blocks